fix regression caused by 52fddc

dzader opened the pull request:
https://github.com/rails/rails/commit/52fddcc653458456f98b3683dffd781cf00b35fe short circuited the call to value.rb::type_cast_from_database (now value.rb::deserialize). In doing so this reintroduced the issue fixed by https://github.com/rails/rails/commit/3525a9b5ebc87bbc9dca4b613b6d3740899190bb

found when investigating https://github.com/rails/rails/issues/18964

Commit:
fix regression caused by 52fddc

matthewd commented:
I'm unclear about how the assumption in 52fddcc653458456f98b3683dffd781cf00b35fe breaks down... surely the underlying DB adapter should never hand us a `true` or `false` for an integer column.

I suspect something's getting confused elsewhere -- so even if this could be made to work, it would be trying to undo damage, instead of avoiding introducing the damage in the first place.

Commit:
only cast_values if the value is not nil

egilburg commented on activerecord/lib/active_record/type/integer.rb:
```diff
@@ -17,8 +17,7 @@ def type
       end
 
       def deserialize(value)
```
Won't the behavior be the same if you just remove this method and let the superclass be called?

Commit:
remove overridden deserialize call

fall back to the default deserialize call which results in cast_values being called

dzader commented:
@matthewd yeah - I agree it does seem strange that we would be getting true/false values for an integer column. I was working under the assumption, based on the previous pull requests/issues, that this was common behavior. Not familiar enough with the PG adapter to know whether that assumption is correct or not

dzader commented:
@rafaelfranca do you have any insight on whether its expected that integer columns can return bool values. From your pull request (https://github.com/rails/rails/commit/3525a9b5ebc87bbc9dca4b613b6d3740899190bb) it seemed that this behavior was common - but I haven't been able to determine if that is true, or whether casting false to 0 was just working around a PG adapter issue.

matthewd commented:
@dzader that's a typecast in the opposite direction: `my_model.integer_column = false`. The database adapter should always return a consistent type.

dzader commented:
@matthewd :+1:  you are correct. totally misread the pull request. closing this and will try and track down the issue in ruby-pg thank you
